server/service/apple_mdm.go

```diff
@@ -2824,3 +2818,144 @@ func (svc *Service) getConfigAppleBMDefaultTeamID(ctx context.Context, appCfg *f
 
 	return tmID, nil
 }
+
+// scepCertRenewalThresholdDays defines the number of days before a SCEP
+// certificate must be renewed.
+const scepCertRenewalThresholdDays = 30
+
+// maxCertsRenewalPerRun specifies the maximum number of certificates to renew
+// in a single cron run.
+//
+// Assuming that the cron runs every hour, we'll enqueue 24,000 renewals per
+// day, and we have room for 24,000 * scepCertRenewalThresholdDays total
+// renewals.
+//
+// For a default of 30 days as a threshold this gives us room for a fleet of
+// 720,000 devices expiring at the same time.
+const maxCertsRenewalPerRun = 100
+
+func RenewSCEPCertificates(
+	ctx context.Context,
+	logger kitlog.Logger,
+	ds fleet.Datastore,
+	config *config.FleetConfig,
+	commander *apple_mdm.MDMAppleCommander,
+) error {
+	if !config.MDM.IsAppleSCEPSet() {
+		logger.Log("inf", "skipping renewal of macOS SCEP certificates as MDM is not fully configured")
+		return nil
+	}
+
+	if commander == nil {
+		logger.Log("inf", "skipping renewal of macOS SCEP certificates as apple_mdm.MDMAppleCommander was not provided")
+		return nil
+	}
+
+	// get a list of certificates that expire within
+	certs, err := ds.GetMDMAppleSCEPCertsCloseToExpiry(ctx, scepCertRenewalThresholdDays, maxCertsRenewalPerRun)
+	if err != nil {
+		return ctxerr.Wrap(ctx, err, "getting certs close to expire")
+	}
+
+	if len(certs) == 0 {
+		logger.Log("info", "no SCEP certificates to renew")
+		return nil
+	}
+
+	// for each certificate, calculate the SHA256 checksum of its DER
+	// contents. This is to match the way nanomdm stores hashes of identity
+	// certificates in the database.
+	//
+	// TODO(roberto): consider pre-calculating these hashes AND/OR adding
+	// efficient indexing.
+	certSHAs := make([]string, len(certs))
+	for i, cert := range certs {
+		parsed, err := apple_mdm.DecodeCertPEM(cert.CertificatePEM)
+		if err != nil {
+			logger.Log("err", "unable to parse SCEP identity certs", "details", err, "cert_serial", cert.Serial)
+			continue
+		}
+		certSHAs[i] = certauth.HashCert(parsed)
+	}
+
+	if len(certSHAs) == 0 {
+		// TODO(code reviewer): should we return an error instead?
```

Comment:
Are you thinking an error would be more visible in the logs?

Reply:
I _think_ the main difference would be that if a job returns an error, then is sent to APM, sentry, etc

I don't know what's the right treatment for this scenario: all expiring hosts have invalid certificates. As I type this: maybe an error, right?

Reply:
So an expired cert never parses and always fails at line 2873?

I think it deserves higher visibility (although if it is running every hour, error alerting  could get noisy)